src/sentry/web/frontend/oauth_token.py

```diff
@@ -42,28 +45,46 @@ def error(self, request: HttpRequest, name, reason=None, status=400):
 
     @never_cache
     def post(self, request: HttpRequest) -> HttpResponse:
+        client_details = self._get_client_details_if_valid_request(request=request)
+
+        if not isinstance(client_details, dict):
+            return client_details  # client_details contains an error response.
+
+        if client_details["grant_type"] == GrantTypes.AUTHORIZATION:
+            return self._get_access_tokens(
+                request=request, application=client_details["application"]
+            )
+        return self._get_refresh_token(request, application=client_details["application"])
+
+    def _get_client_details_if_valid_request(self, request: Request) -> Union[dict, HttpResponse]:
```

Comment:
Can we maintain a single response type?

these types are pretty far off. If we want to return an error response - then lets do that up in the `def post` and simply return a dict type with the relevant information to make that decision